src/git-merge-pr.rs

```diff
@@ -4,18 +4,171 @@ extern crate git_rsl;
+    let ws = Workspace::new(&mut repo)?;
+
+    // secure-fetch src, dest
+    let mut remote = ws.repo.find_remote("origin")
+        .chain_err(|| "unable to find remote 'origin'")?;
+
+    // secure fetch both branches from origin
+    secure_fetch(&ws.repo,
+                 &mut remote,
+                 &vec![opts.branch_to_merge.as_str(),
+                       opts.destination_branch.as_str()])?;
+
+
+    let origin_destination_branch = format!("refs/remotes/origin/{}", opts.destination_branch);
+    let origin_branch_to_merge = format!("refs/remotes/origin/{}", opts.branch_to_merge);
+
+    let destination_branch_oid = ws.repo.refname_to_id(&origin_destination_branch)?;
+    let branch_to_merge_oid = ws.repo.refname_to_id(&origin_branch_to_merge)?;
 
-    println!("merge-pr");
+    let destination_branch_commit = &ws.repo.find_commit(destination_branch_oid)?;
+    let branch_to_merge_commit = &ws.repo.find_commit(branch_to_merge_oid)?;
+
+    let destination_branch_annotated_commit = &ws.repo.find_annotated_commit(destination_branch_oid)?;
+    let branch_to_merge_annotated_commit = &ws.repo.find_annotated_commit(branch_to_merge_oid)?;
+
+    // The original shell version runs 'git gc' here; I'm omitting it because it
+    // gc is a potentially destructive operation, and doing it automatically is
+    // somewhat dangerous.
+
+    // Check out the local destination_branch and ff to match the remote, if
+    // possible.
+    //
+    // The shell version did a series of machinations involving a detached HEAD;
+    // that won't work here because git-rsl's implementation of secure-push can
+    // only push a branch to a remote version of the same name (i.e. you can't
+    // push HEAD:branchname, only branchname:branchname)
+    ws.repo.set_head(&local_destination_branch)?;
+
+    // secure-fetch leaves the working copy in a crazy state, with the RSL
+    // branch checked out, so we need to force checkout the destination branch
+    ws.repo.checkout_head(Some(git2::build::CheckoutBuilder::new().force()))?;
+
+    match ws.repo.merge_analysis(&[&destination_branch_annotated_commit])? {
+        (git2::MergeAnalysis::ANALYSIS_UP_TO_DATE, _) =>
+            println!("local {} branch is up to date", opts.destination_branch),
+        (git2::MergeAnalysis::ANALYSIS_FASTFORWARD, _) => {
+            println!("fast-forwarding {} to {}", opts.destination_branch,
+                     destination_branch_oid);
+            git::fast_forward_onto_head(&ws.repo, &origin_destination_branch)?
+        },
+        _ =>
+            bail!("your local {} branch is not in the history of origin/{},\
+                   so we can't do a fast-forward merge. You've probably made \
+                   a local commit to it. Reconcile the situation (probably by \
+                   resetting to origin/{}) and try again.",
+                  opts.destination_branch,
+                  opts.destination_branch,
+                  opts.destination_branch)
+    }
```

Comment:
You can put a number inside the braces that references the index of the argument ex. `"your local {0} branch ...`. Then you only have to list the argument once.

Reply:
woah, that's a cool feature.